### PR TITLE
General cleanup, add Win32 support, and add Linux-to-Win32 cross-compilation

### PR DESCRIPTION
**Pull request by rgprog**
This PR adds Win32 support to `capnp.cc`, along with the (extensive!) build infrastructure required to make it available through NPM. See this PR's individual commit messages for the full story.

**As written, this PR will break Rewind**, by (1) changing the CLI commands used for cross-compilation, and (2) refactoring `build.js` and `build-capnp.sh` without testing those changes for any Darwin builds. I can see two options:

- Update Rewind to use the new CLI commands, and then test the various Rewind workflows which produce a Darwin build. Some of the possible failures are quite late-bound in my experience (e.g. failing to dynamically load `capnp.node`), so this would probably require running the resulting builds on an actual Mac.
- Pin Rewind's `capnp` dependency against the most recent commit before this PR (c63332b).

### CODE REVIEW

File: build-capnp.bat
Comment (p--b) [inline build-capnp.bat:1]: As a general principle, I wonder if it would be better for us to be defaulting to powershell scripts rather than batch files - it's a much more pleasant language to work with

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]: As `end` doesn't do any cleanup, `return total;` would be clearer here and below.

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]:
```suggestion
      int bytes_written = write(fd, buf, len);
```

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]: Prefer using `#if 0` rather than commenting out code, ideally with a comment explaining why

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc:3134]: As above

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc:371]: This seems to be a substantive change from the original that is not win32 only - a comment explaining the reason for the divergence from upstream would be useful, if not putting this in its own commit.

File: build.js
Comment (p--b) [inline build.js]: Logging that we're reusing a precompiled binary might be a helpful pointer if someone is debugging things in future.

File: README.md
Comment (p--b) [inline README.md]: Moving the readmes around like this is likely to cause much faff when attempting to merge changes from upstream, hence why our own readme was previously added  as a separate file. By all means we can add a pointer to our readme to the upstream readme, but I don't think we should move them around like this.

File: package.json
Comment (p--b) [inline package.json:26]: I don't see anything that causes the new toolchain to be run with the replacement for `patch_path` (`patch-lib-path`) which will break the darwin build.

**Review by p--b — APPROVED** [review]
Given that we're planning on shipping a Darwin client in the first instance, I think we should ensure that these changes result in a functional Darwin module before merging.

Separately, for Rewind, I'm happy for us to pin to an older commit for now rather than adapting its toolchain for this new version; a PR doing that should occur before merging this PR.

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]: As per offline discussion.

**Review by p--b — CHANGES_REQUESTED** [review]
As per offline discussion

**Comment by rgprog** [comment]
> Given that we're planning on shipping a Darwin client in the first instance, I think we should ensure that these changes result in a functional Darwin module before merging.

No worries - what's our plan for testing that? I have no access to a Mac.

File: build-capnp.bat
Comment (rgprog) [reply build-capnp.bat:1]: Moved this one to an offline discussion, for better visibility

File: src/node-capnp/capnp.cc
Comment (rgprog) [reply src/node-capnp/capnp.cc:371]: Sorry, I'm confused - this is already in its own commit, 80ebee24ce0db73e3b06440d68325e3f7b43259a

(Looks like the commit message is back-to-front! I'll reword it now.)

File: README.md
Comment (rgprog) [reply README.md]: It took several months for me to realise that `README-FOURIER.md` existed at all, which is why I leaned hard towards maximum visibility here - I'd expect our readme to be read much more often than it's edited.

That being said, it looks like Git's rename detection has already failed us here, so editing a link into the original readme does sound like a pretty fair compromise.

File: package.json
Comment (rgprog) [reply package.json:26]: Good catch - solved this by hardcoding the `patchLibPath`, since the build paths themselves are currently hardcoded anyway.

**Comment by p--b** [comment]
> what's our plan for testing that

Henry should be able to run the skeletal client that we've built on his laptop - it won't have anything to connect to, but even getting a connection failed error message is a proof of life

**Comment by p--b** [comment]
Unfortunately this branch also currently results in the client hanging on Linux, seeming to jam the event loop after a request is dispatched on a capability. I have narrowed this down to the source changes to `capnp.cc` (i.e. it doesn't seem to be caused by the build system changes) - checking out `capnp.cc` from main and rebuilding produces a functioning module.

Update: I thought the culprit is the change `uv_poll_init -> uv_poll_init_socket`, but I was mistaken. It seems to be caused by the code to avoid writev...

File: src/node-capnp/capnp.cc
Comment (p--b) [reply src/node-capnp/capnp.cc:371]: Ah, I had missed that it was in its own commit already, apologies. I suppose my point is that given the number commits in this branch currently I was assuming that we would squash it into main, at which point we would lose the reasoning for this change. (TBF We could choose to rebase it in as the commits are all useful, but the PR description doesn't mention that.)

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]: This change seems to cause Linux (and potentially macos) clients to hang their event loop as soon as a capability is used to dispatch an RPC request.

Comment (p--b) [reply src/node-capnp/capnp.cc]: (Applying [this commit](https://github.com/fourieraudio/node-capnp/commit/d6580a7009e39756a247b3a8270311c909a79e6c), reverting the change, allows the client to run normally on Linux)

File: src/node-capnp/capnp.cc
Comment (rgprog) [reply src/node-capnp/capnp.cc:371]: The plan is to rebase

File: src/node-capnp/capnp.cc
Comment (rgprog) [reply src/node-capnp/capnp.cc]: @p--b: On the current `win32-cross` branch, would you expect this bug to reproduce if I `cargo run --release` Director, start up the new Transform frontend (the `client-infra` branch of Transform), and connect to `127.0.0.1`?

If so, this bug isn't reproducing on Windows. I don't have a Linux test-case which I can use to investigate that side of it (because our frontend doesn't run on WSL, seemingly due to filesystem permission errors). I could try to fix this by staring at it and thinking hard, but the first five minutes of that haven't led to any epiphanies, except that the `writeResult < 0` comparison looks a bit suspicious.

Do you want to look into this, or should we simply `#ifdef` it and pray for forgiveness later?

Comment (p--b) [reply src/node-capnp/capnp.cc]: Yes, that ought to be enough. I must confess I also can't see what's wrong by inspection. At this stage I think let's go ifdef; writev is possibly marginally preferable anyway for perf.

Comment (rgprog) [reply src/node-capnp/capnp.cc]: This should now be patched

**Comment by rgprog** [comment]
Unless there's any urgency, I'd like to hold off on merging this PR until I've made some progress on TF-361 and TF-362; I have a feeling that both tickets are likely to expose additional bugs.

**Comment by henryharrod** [comment]
Any update on this PR? Got mine dependent on it (in order to build transform desktop)

**Comment by rgprog** [comment]
TF-362 was finished (pending review) without exposing any new bugs. TF-361 is still pending, but perhaps a little less relevant to this PR, because it should bear a pretty close resemblance to our existing frontend capnproto code.

No objections to merging now (by rebase), if it would make your life easier! It's just blocked on an approving code review from @p--b.

File: src/node-capnp/capnp.cc
Comment (p--b) [inline src/node-capnp/capnp.cc]: nitpick: given that we might upstream this, probably worth not removing this

File: build.js
Comment (p--b) [inline build.js:155]: Is it worth tagging on an `else { throw new Error("Unsupported platform") }` in case we try to add building from `darwin` in the future (I know it's tested before above but makes the change easier to make in future)

File: build.js
Comment (p--b) [inline build.js]:
```suggestion
```

File: build.js
Comment (p--b) [inline build.js:332]: Nitpick: changes like this one (all diff :327->:357 of `patchLibs`) don't seem to add anything to the functionality but do clutter the history & increase the size of the diff. Unless this change was made automatically by a formatter, consider avoiding changes like this to make it easier to trace the history of the code & reduce review size - it's really not clear from this diff that the only substantive change to this function is on :325.

File: binding.gyp
Comment (p--b) [inline binding.gyp:36]: Why are we adding `--whole-archive`? Deserves a comment

File: binding.gyp
Comment (p--b) [inline binding.gyp:42]: This seems odd, given :34. If it's correct, deserves a comment

File: build-capnp.sh
Comment (p--b) [inline build-capnp.sh:68]: Nitpick: this is a substantive change: previously on linux we linked against the system's libcapnp. Whilst I'm not against this in principle for reproducibility reasons, I would have expected this to have been called out more visibly in this PR.

File: build-capnp.sh
Comment (p--b) [inline build-capnp.sh]: Please check if this already exists and reuse it if it does. Redownloading stuff every time we do a build is wasteful, adds up fast, and makes developing offline / remotely difficult.

File: build-capnp.sh
Comment (p--b) [inline build-capnp.sh]: Now that we are not using the system's libcapnp, `install-deps.sh` in this repository is unnecessary and should be removed.

**Comment by henryharrod** [comment]
That would be great - thanks @rgprog. It's currently holding up TF-13 as that depends on this main branch (node-capnp) being up to speed with the changes in win32-cross-revert-writev (which is itself dependent on this PR merging). 

If you could do this week that would be great, as then can try to merge TF-13 by the time I go away next Weds.

Ta1

**Review by p--b — APPROVED** [review]
Looks good pending comments.

As a nitpick, this PR has a lot of commits which aren't really the sort of commits I'd want to see in the history (e.g. 5b0a55dc, 41941fa9, dc0c6e61, 85005449, f7f699f, 1e74538). If we were squashing this branch it would be irrelevant, but given we plan to rebase, having commits like those in the history make it much more difficult to helpfully use `git log` / `git blame` to understand when, how (and critically _why_) changes were made to the software.

In the interests of time we might just want to merge this (as the time we've spent on enabling Windows development is already significantly above what it would have taken to just run a VM, and this work is not user visible in the short term), but in general (and especially for work that we might consider upstreaming), our future selves will thank us for not making a mess of the repository history.

File: binding.gyp
Comment (p--b) [reply binding.gyp:36]: I know what it does, I don't know _why_ we're using it here: I don't see why we should be shipping bits of libkj we're not using

Comment (rgprog) [reply binding.gyp:36]: I believe introducing the flag resolved some stubborn "undefined symbol" errors, but it's been nearly two months, so I can't recall the fine details. I don't think I investigated whether the flag could be omitted for individual static libraries like `libkj`.

Properly reevaluating whether the `--whole-archive` flag is necessary here would take experimentation. If you'd like me to spend a bit of time on that, let me know.

**Pull request merged**
